pandas/io/orc.py

```diff
@@ -52,3 +57,64 @@ def read_orc(
     with get_handle(path, "rb", is_text=False) as handles:
         orc_file = orc.ORCFile(handles.handle)
         return orc_file.read(columns=columns, **kwargs).to_pandas()
+
+
+def to_orc(
+    df: DataFrame,
+    path: FilePath | WriteBuffer[bytes] | None = None,
+    engine: Literal["pyarrow"] = "pyarrow",
+    index: bool = None,
+    **kwargs,
+) -> bytes:
+    """
+    Write a DataFrame to the ORC format.
+    Parameters
+    ----------
+    df : DataFrame
+    path : str or file-like object, default None
+        If a string, it will be used as Root Directory path
+        when writing a partitioned dataset. By file-like object,
+        we refer to objects with a write() method, such as a file handle
+        (e.g. via builtin open function). If path is None,
+        a bytes object is returned. Note that currently the pyarrow
+        engine doesn't work with io.BytesIO.
+    engine : {{'pyarrow'}}, default 'pyarrow'
+        Parquet library to use, or library it self, checked with 'pyarrow' name
+        and version >= 5.0.0
+    index : bool, default None
+        If ``True``, include the dataframe's index(es) in the file output. If
+        ``False``, they will not be written to the file.
+        If ``None``, similar to ``infer`` the dataframe's index(es)
+        will be saved. However, instead of being saved as values,
+        the RangeIndex will be stored as a range in the metadata so it
+        doesn't require much space and is faster. Other indexes will
+        be included as columns in the file output.
+    kwargs
+        Additional keyword arguments passed to the engine
+    Returns
+    -------
+    bytes if no path argument is provided else None
+    """
+    if index is None:
+        index = df.index.names[0] is not None
+
+    if engine != "pyarrow":
+        raise ValueError(f"engine must be 'pyarrow'")
+    engine = import_optional_dependency(engine, min_version="5.0.0")
+
+    if hasattr(path, "write"):
+        engine.orc.write_table(
+            engine.Table.from_pandas(df, preserve_index=index), path, **kwargs
+        )
+    else:
+        # to bytes: pyarrow auto closes buffers hence we read a pyarrow buffer
```

Comment:
Can you add a test that calls write_table directly with BytesIO or a different handle to then assert that it is closed after pyarrow wrote the dataframe? Then we will automatically know when we can remove the BufferOutputStream.

Reply:
@twoertwein https://github.com/apache/arrow/pull/12231 I just got the BytesIO issue fixed in Arrow. Since 7.0.0 will be released very soon I'm not sure whether this one will be added.

Reply:
@twoertwein  That was actually fixed in 7.0.0 so we are good to go now.